### ISSUE TITLE
Motion estimation needs a lot of memory

### ISSUE THREAD

**Issue by jdickerson95** [teamtomo/torch-motion-correction#2]
The motion estimation uses a lot of memory in this step:
[

    patches = images[..., patch_idx_d, patch_idx_h, patch_idx_w]](https://github.com/teamtomo/torch-motion-correction/blob/main/src/torch_motion_correction/patch_grid/_patch_grid.py#L121)

Makes running on a GPU in particular difficult.
Not sure what the best fix is / if it needs fixing, but a lot of people could run into this.

**Comment by alisterburt** [comment]
thought with claude for a couple of mins but haven't tried this out

```python
import torch
import einops
from typing import Tuple, Union, Sequence
import numpy as np

class LazyPatchGrid3D:
    """
    A lazy patch grid that computes patches on-demand to avoid memory overhead.
    Presents the same interface as a pre-computed patch grid but only extracts
    patches when they are actually accessed.
    """
    
    def __init__(
        self,
        images: torch.Tensor,
        patch_shape: Tuple[int, int, int],
        patch_step: Tuple[int, int, int],
        distribute_patches: bool = True,
    ):
        """
        Parameters
        ----------
        images: torch.Tensor
            (..., d, h, w) array of 3D images.
        patch_shape: Tuple[int, int, int]
            (patch_d, patch_h, patch_w) of patches to be extracted.
        patch_step: Tuple[int, int, int]
            The target distance between patch centers in dimensions d, h, and w.
        distribute_patches: bool
            Whether to distribute patches across the entire dimension length.
        """
        self.images = images
        self.patch_shape = patch_shape
        self.patch_step = patch_step
        self.distribute_patches = distribute_patches
        
        # Pre-compute patch centers and indices - these are lightweight
        self.patch_centers = patch_grid_centers(
            image_shape=images.shape[-3:],
            patch_shape=patch_shape,
            patch_step=patch_step,
            distribute_patches=distribute_patches,
            device=images.device,
        )
        
        self.patch_idx_d, self.patch_idx_h, self.patch_idx_w = patch_grid_indices(
            image_shape=images.shape[-3:],
            patch_shape=patch_shape,
            patch_step=patch_step,
            distribute_patches=distribute_patches,
            device=images.device,
        )
        
        # Store grid dimensions for shape inference
        self.grid_shape = self.patch_centers.shape[:-1]  # (grid_d, grid_h, grid_w)
        
        # Full shape would be (..., grid_d, grid_h, grid_w, patch_d, patch_h, patch_w)
        self._shape = images.shape[:-3] + self.grid_shape + patch_shape
    
    @property
    def shape(self) -> torch.Size:
        """Return the shape that the full patch grid would have."""
        return torch.Size(self._shape)
    
    @property
    def device(self) -> torch.device:
        """Return the device of the underlying images."""
        return self.images.device
    
    @property
    def dtype(self) -> torch.dtype:
        """Return the dtype of the underlying images."""
        return self.images.dtype
    
    def __getitem__(self, key) -> torch.Tensor:
        """
        Extract patches on-demand based on indexing.
        
        This is where the magic happens - only the requested patches are computed.
        """
        # Handle the case where we're indexing into the patch grid dimensions
        if isinstance(key, tuple) and len(key) >= 2:
            # Extract the indices for the grid dimensions
            batch_idx = key[:-3] if len(key) > 3 else ()
            grid_idx = key[-3:] if len(key) >= 3 else key
            
            # Handle ellipsis and slices properly
            if len(grid_idx) == 3:
                gd_idx, gh_idx, gw_idx = grid_idx
            elif len(grid_idx) == 2:
                gh_idx, gw_idx = grid_idx
                gd_idx = slice(None)
            else:
                raise IndexError("Invalid indexing for patch grid")
            
            # Get the subset of patch indices we need
            patch_idx_d_subset = self.patch_idx_d[gd_idx]
            patch_idx_h_subset = self.patch_idx_h[gh_idx]  
            patch_idx_w_subset = self.patch_idx_w[gw_idx]
            
            # Select the corresponding image regions
            if batch_idx:
                image_subset = self.images[batch_idx]
            else:
                image_subset = self.images
                
            # Extract only the patches we need
            patches = image_subset[..., patch_idx_d_subset, patch_idx_h_subset, patch_idx_w_subset]
            
            return patches
        
        else:
            # For simple indexing, fall back to extracting specific grid positions
            if isinstance(key, (int, slice)):
                # Single dimension indexing - assume it's the first grid dimension
                return self._extract_patches_at_grid_positions(key, slice(None), slice(None))
            else:
                raise IndexError(f"Unsupported indexing pattern: {key}")
    
    def _extract_patches_at_grid_positions(self, gd_idx, gh_idx, gw_idx) -> torch.Tensor:
        """Extract patches at specific grid positions."""
        patch_idx_d_subset = self.patch_idx_d[gd_idx]
        patch_idx_h_subset = self.patch_idx_h[gh_idx]
        patch_idx_w_subset = self.patch_idx_w[gw_idx]
        
        patches = self.images[..., patch_idx_d_subset, patch_idx_h_subset, patch_idx_w_subset]
        return patches
    
    def get_patch_centers_subset(self, gd_idx, gh_idx, gw_idx) -> torch.Tensor:
        """Get patch centers for a subset of the grid."""
        return self.patch_centers[gd_idx, gh_idx, gw_idx]
    
    def random_subset(self, n_patches: int) -> Tuple[torch.Tensor, torch.Tensor]:
        """
        Get a random subset of patches - optimized version of your use case.
        
        Parameters
        ----------
        n_patches: int
            Number of random patches to extract
            
        Returns
        -------
        patches: torch.Tensor
            Random subset of patches
        centers: torch.Tensor
            Corresponding patch centers
        """
        # Generate random indices
        gd_max, gh_max, gw_max = self.grid_shape
        
        patch_subset_idx = torch.randint(
            low=0, 
            high=torch.tensor([gd_max, gh_max, gw_max], device=self.device),
            size=(n_patches, 3),
            device=self.device
        )
        
        idx_gd, idx_gh, idx_gw = patch_subset_idx.t()
        
        # Extract only the requested patches
        patches = self[:, idx_gd, idx_gh, idx_gw]
        centers = self.patch_centers[idx_gd, idx_gh, idx_gw]
        
        return patches, centers
    
    def __repr__(self) -> str:
        return (f"LazyPatchGrid3D(shape={self.shape}, "
                f"patch_shape={self.patch_shape}, "
                f"grid_shape={self.grid_shape})")


def patch_grid_3d_lazy(
    images: torch.Tensor,
    patch_shape: Tuple[int, int, int],
    patch_step: Tuple[int, int, int],
    distribute_patches: bool = True,
) -> Tuple[LazyPatchGrid3D, torch.Tensor]:
    """
    Create a lazy patch grid that extracts patches on-demand.
    
    Returns
    -------
    lazy_patches: LazyPatchGrid3D
        Lazy patch grid object
    patch_centers: torch.Tensor
        Pre-computed patch centers
    """
    lazy_grid = LazyPatchGrid3D(
        images=images,
        patch_shape=patch_shape,
        patch_step=patch_step,
        distribute_patches=distribute_patches,
    )
    
    return lazy_grid, lazy_grid.patch_centers


# Example usage demonstrating memory efficiency
if __name__ == "__main__":
    # Create some dummy 3D images
    batch_size, d, h, w = 2, 64, 128, 128
    images = torch.randn(batch_size, d, h, w)
    
    patch_shape = (16, 32, 32)
    patch_step = (8, 16, 16)
    
    # Create lazy patch grid
    lazy_patches, patch_centers = patch_grid_3d_lazy(
        images, patch_shape, patch_step
    )
    
    print(f"Image shape: {images.shape}")
    print(f"Lazy patch grid shape: {lazy_patches.shape}")
    print(f"Memory usage: Only centers computed upfront")
    
    # Your optimization loop example
    n_iterations = 100
    n_patches_per_batch = 32
    gd, gh, gw = lazy_patches.grid_shape
    
    for i in range(n_iterations):
        # Method 1: Using your existing approach with lazy evaluation
        patch_subset_idx = np.random.randint(
            low=(0, 0, 0), high=(gd, gh, gw), size=(n_patches_per_batch, 3)
        )
        idx_gd, idx_gh, idx_gw = einops.rearrange(patch_subset_idx, 'b idx -> idx b')
        
        # This only computes the patches we actually need!
        patch_subset = lazy_patches[:, idx_gd, idx_gh, idx_gw]
        patch_subset_centers = lazy_patches.get_patch_centers_subset(idx_gd, idx_gh, idx_gw)
        
        # Method 2: Using the built-in random_subset method
        # patches, centers = lazy_patches.random_subset(n_patches_per_batch)
        
        print(f"Iteration {i}: extracted {patch_subset.shape} patches")
        
        if i >= 2:  # Just show a few iterations
            break
```

**Comment by jdickerson95** [comment]
Nice! This is along the lines of what I was thinking. I can give it a try soon.

I was also having memory issues with correct motion. This seemed to be to do with the gradients (it was worse the longer the movie), so it could be solved by turning the gradients off for correct motion. I was thinking it could be an option, with the default being no gradients for correct motion.

**Comment by alisterburt** [comment]
eep - if switching off gradients solves it that could be great

If performance for correct motion is an issue I remember I used to have an optimization for that where I would evaluate a ~5-10x oversampled grid with the B-spline interpolation here
https://github.com/teamtomo/torch-motion-correction/blob/c5293916f0d67fce6087daf2b64ebeaf53be1c02/src/torch_motion_correction/evaluate_deformation_grid.py#L5-L12

then evaluate that oversampled grid in 2D with `F.grid_sample` in bicubic

**Comment by alisterburt** [comment]
can't find the code right now, probably on an old hard drive and never got pushed
